src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs

```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net;
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Dcp;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+
+namespace Aspire.Hosting.Dashboard;
+
+/// <summary>
+/// Hosts a gRPC service (via <see cref="DashboardService"/>) that a dashboard can connect to.
+/// Configures DI and networking options for the service.
+/// </summary>
+internal sealed class DashboardServiceHost : IHostedService
+{
+    private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
+    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+
+    /// <summary>
+    /// <see langword="null"/> if <see cref="DistributedApplicationOptions.DashboardEnabled"/> is <see langword="false"/>.
+    /// </summary>
+    private readonly WebApplication? _app;
+
+    public DashboardServiceHost(
+        DistributedApplicationOptions options,
+        DistributedApplicationModel applicationModel,
+        KubernetesService kubernetesService)
+    {
+        if (!options.DashboardEnabled)
+        {
+            return;
+        }
+
+        var builder = WebApplication.CreateBuilder();
+
+        builder.Services.AddGrpc();
+        builder.Services.AddSingleton(applicationModel);
+        builder.Services.AddSingleton(kubernetesService);
+        builder.Services.AddSingleton<DashboardServiceData>();
+
+        builder.WebHost.ConfigureKestrel(ConfigureKestrel);
+
+        _app = builder.Build();
+
+        _app.MapGrpcService<DashboardService>();
+        _app.MapGet("/", () => "Communication with gRPC endpoints must be made through a gRPC client. To learn how to create a client, visit: https://go.microsoft.com/fwlink/?linkid=2086909");
+
+        return;
+
+        static void ConfigureKestrel(KestrelServerOptions kestrelOptions)
+        {
+            var uris = GetAddressUris(DashboardServiceUrlVariableName, DashboardServiceUrlDefaultValue);
+
+            foreach (var uri in uris)
+            {
+                if (uri.IsLoopback)
+                {
+                    kestrelOptions.ListenLocalhost(uri.Port, ConfigureListen);
+                }
+                else
+                {
+                    kestrelOptions.Listen(IPAddress.Parse(uri.Host), uri.Port, ConfigureListen);
+                }
+
+                void ConfigureListen(ListenOptions options)
+                {
+                    // Force HTTP/2 for gRPC, so that it works over non-TLS connections
+                    // which cannot negotiate between HTTP/1.1 and HTTP/2.
+                    options.Protocols = HttpProtocols.Http2;
+
+                    if (string.Equals(uri.Scheme, "https", StringComparison.Ordinal))
```

Comment:
Ignore case? Someone could pass in uppercase strings in the env variable.

(or does Uri normalize scheme to lowercase?)

Reply:
This code exists in a few places today, and I copied it from somewhere. Please file a separate issue if it is actually an issue.